### PR TITLE
Prevent duplicate tag names

### PR DESCRIPTION
**Pull request by jakobroehrl**
* Resolves: #586 
* Target version: master 

### Summary
entering a duplicate title is not possible

### TODO

- [ ] ...

### Checklist

- [ ] Code is properly formatted
- [ ] Sign-off message is added to all commits
- [ ] Tests (unit, integration, api and/or acceptance) are included
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by juliushaertl** [comment]
Thank you very much for your contribution @jakobroehrl. Can you also add this to the update method for labels? Otherwise creating duplicate labels by renaming them is still possible. Maybe you can move the check to a separate private method, so we don't have duplicate code.

**Comment by jakobroehrl** [comment]
@juliushaertl
Thanks for your advice. I added a new function and refered it from update and create.

- My check in the update function has a problem: In the UI remains the duplicate title until reloading the page. Do you have an idea where I can fix that behaviour?

- visual indicator in the UI: The visual indicator for the other label checks (title === false, color === false, ...) is missing, too. I would love to implement it. Are there any code snips oder examples I can use?

Thanks and Greetings!

**Comment by juliushaertl** [comment]
You should be able to add a second callback to the then call here, to catch the error when sending the create request to the server:
https://github.com/nextcloud/deck/blob/master/js/controller/BoardController.js#L330

Although it isn't nice, i would just use `OC.Notification.showTemporary('Duplicate label name is not allowed');` to render a notification to the user, since we are currently reworking our UI anyway.

**Comment by juliushaertl** [comment]
> How can I distinguish the different error messages in the BoardController? I get the same message on every error. BoardController.js line 335

This seems to be an issue with the current way our service layer works. The generic error message is coming from https://github.com/nextcloud/deck/blob/master/js/service/ApiService.js#L122 so we would need to return the proper error from the server here.

> If the label update fails I want to set the value from before. Where can I get it 
from? In $scope and label is only the new value. BoardController.js line 343

You need to add variable to store the previous value during the editing process, similar to what we do when editing the board title: https://github.com/nextcloud/deck/blob/master/js/controller/BoardController.js#L47

**Comment by jakobroehrl** [comment]
@juliushaertl: I hope, it's ready to merge now :)

**Comment by juliushaertl** [comment]
@jakobroehrl I rebased this to let the tests run against current master. Any change you could look into the failing unit tests:

```

There were 2 errors:
1) OCA\Deck\Service\LabelServiceTest::testCreate
Invalid argument supplied for foreach()
/home/travis/build/nextcloud/server/apps/deck/lib/Service/LabelService.php:97
/home/travis/build/nextcloud/server/apps/deck/tests/unit/Service/LabelServiceTest.php:74
2) OCA\Deck\Service\LabelServiceTest::testUpdate
Invalid argument supplied for foreach()
/home/travis/build/nextcloud/server/apps/deck/lib/Service/LabelService.php:170
/home/travis/build/nextcloud/server/apps/deck/tests/unit/Service/LabelServiceTest.php:95
ERRORS!
```

File: lib/Service/LabelService.php [hunk 1]
```diff
@@ -166,6 +167,7 @@ public function update($id, $title, $color) {
 		$label = $this->find($id);
 		//$this->checkDuplicateTitle($label->getBoardId(), $title);
 		
+		$boardLabels = array();
```
Comment (juliushaertl) [inline lib/Service/LabelService.php]: you will need to create a mock in the tests for the findAll method on labelMapper, so that this either returns an empty array or better (to properly test the code) contain an actual list of labels that will then be used in the foreach loop during the test.

File: lib/Service/LabelService.php [hunk 2]
```diff
@@ -91,6 +91,18 @@ public function create($title, $color, $boardId) {
 		}
 
 		$this->permissionService->checkPermission(null, $boardId, Acl::PERMISSION_MANAGE);
+
+		//$this->checkDuplicateTitle($boardId, $title);
+		$boardLabels = $this->labelMapper->findAll($boardId);
+		if (is_array($boardLabels) || is_object($boardLabels)) {
```
Comment (juliushaertl) [inline lib/Service/LabelService.php]: No need for is_object, $boardLabels will either be an array or null

File: lib/Service/LabelService.php [hunk 3]
```diff
@@ -91,6 +91,18 @@ public function create($title, $color, $boardId) {
 		}
 
 		$this->permissionService->checkPermission(null, $boardId, Acl::PERMISSION_MANAGE);
+
+		//$this->checkDuplicateTitle($boardId, $title);
```
Comment (juliushaertl) [inline lib/Service/LabelService.php]: Could you still move this to a separate method. Then we don't have duplicate code.

**Comment by juliushaertl** [comment]
@jakobroehrl I rebase your changes and pushed a small commit to fix the codacy warnings. Let's get this in :+1:
